src/deps-index.ts

```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
+  });
+
+  visited.add(url);
+
+  // Every unvisited direct eager dependency's lazy dependencies are recognized
+  // as transitively lazy unless already recognized as eager.  All others are
+  // treated as eager.
+  Array.from(map.eager).filter((d) => !visited.has(d)).forEach((d) => {
+    const deps = getTransitiveDependencies(d, analysis, visited);
+    deps.eager.forEach((dep) => map.eager.add(dep) && map.lazy.delete(dep));
+    deps.lazy.forEach((dep) => map.eager.has(dep) || map.lazy.add(dep));
+    deps.lazyEdges.forEach((dep) => map.lazyEdges.add(dep));
+  });
+
+  // Every unvisited direct lazy dependency's dependencies are recognized
+  // as transitively lazy unless already defined as eager.
+  Array.from(map.lazy).filter((d) => !visited.has(d)).forEach((d) => {
+    const deps = getTransitiveDependencies(d, analysis, visited);
+    deps.eager.forEach(
+        (dep) => map.eager.has(dep) ? map.lazy.delete(dep) : map.lazy.add(dep));
+    deps.lazy.forEach((dep) => map.eager.has(dep) || map.lazy.add(dep));
+    deps.lazyEdges.forEach((dep) => map.lazyEdges.add(dep));
+  });
 
+  visited.delete(url);
+  return map;
+}
+
+/**
+ * Analyzes all entrypoints and determines each of their transitive
+ * dependencies.
+ * @param entrypoints Urls of entrypoints to analyze.
+ * @param analyzer
+ * @return a dependency index of every entrypoint, including entrypoints that
+ *     were discovered as lazy edges in the graph.
+ */
 export async function buildDepsIndex(
     entrypoints: UrlString[], analyzer: Analyzer): Promise<DepsIndex> {
-  const entrypointToDependencies: Map<UrlString, Set<UrlString>> = new Map();
-  const dependenciesToEntrypoints: Map<UrlString, Set<UrlString>> = new Map();
-  const queue = Array.from(entrypoints);
-  const visitedEntrypoints = new Set<UrlString>();
-  while (queue.length > 0) {
-    const entrypoint = queue.shift()!;
-    if (visitedEntrypoints.has(entrypoint)) {
-      continue;
-    }
-    const dependencyEntry =
-        await _getTransitiveDependencies(entrypoint, entrypoints, analyzer);
-    const dependencies = new Set(dependencyEntry.eager);
-    dependencies.add(entrypoint);
-    entrypointToDependencies.set(entrypoint, dependencies);
-    for (const lazyDependency of dependencyEntry.lazy.values()) {
-      if (!visitedEntrypoints.has(lazyDependency)) {
-        queue.push(lazyDependency);
-      }
-    }
-  }
-  entrypointToDependencies.forEach((dependencies, entrypoint, map) => {
-    for (const dependency of dependencies) {
-      if (!dependenciesToEntrypoints.has(dependency)) {
-        dependenciesToEntrypoints.set(dependency, new Set());
-      }
-      const entrypointSet = dependenciesToEntrypoints.get(dependency)!;
-      entrypointSet.add(entrypoint);
-    }
+  const depsIndex = {entrypointToDeps: new Map<UrlString, Set<UrlString>>()};
+  const analysis = await analyzer.analyze(entrypoints);
+  const processed = new Set<UrlString>();
+  const unprocessed = new Set<UrlString>(entrypoints);
```

Comment:
Because it's a set, if you only added entries to `unprocessed` and never deleted then you could get rid of `processed` entirely. Could rename it from `unprocessed` to just `urls`